subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,426 @@
+	__ASSERT(sensor_api, "init sensor, sensor device sensor_api is NULL");
+
+	/* physical sensor has no reporter, conns_num is 0 */
+	if (conns_num == 0) {
+		sensor->conns = NULL;
+	}
+
+	for (i = 0; i < conns_num; i++) {
+		conn = &sensor->conns[i];
+		reporter = get_reporter_sensor(ctx, sensor, i);
+		__ASSERT(reporter, "sensor's reporter should not be NULL");
+		ret |= init_each_connection(ctx, conn, reporter, sensor);
+		if (ret) {
+			LOG_ERR("init sensor, init_each_connection error:%d", ret);
+			return ret;
+		}
+
+		/* device tree required sensor connection between reporter and client cannot be
+		 * opened or closed any more, so it is called fixed connection.
+		 */
+		conn->index = ctx->fixed_connection_count++;
+		LOG_DBG("init sensor, reporter:%s, client:%s, connection:%d",
+			reporter->dev->name, sensor->dev->name, conn->index);
+
+		__ASSERT(conn->index < CONFIG_SENSING_MAX_CONNECTION_COUNT,
+			 "sensor connection number:%d exceed %d",
+			 conn->index, CONFIG_SENSING_MAX_CONNECTION_COUNT);
+
+		sensor->conns = conn;
+		ctx->conns[conn->index] = conn;
+		/* link connection to its repoter's client_list */
+		sys_slist_append(&reporter->client_list, &conn->snode);
+
+		conn_idx[i] = conn->index;
+	}
+
+	/* physical sensor is working at polling mode by default,
+	 * virtual sensor working mode is inherited from its reporter
+	 */
+	if (is_phy_sensor(sensor)) {
+		sensor->mode = SENSOR_TRIGGER_MODE_POLLING;
+	}
+
+	return sensor_api->init(sensor->dev, &sensor->dt.info, conn_idx, conns_num);
+}
+
+/* create struct sensing_sensor *sensor according to sensor device tree */
+static struct sensing_sensor *create_sensor(const struct sensing_dt_info *dt)
+{
+	struct sensing_sensor_ctx *sensor_ctx;
+	struct sensing_sensor *sensor;
+	uint16_t sample_size, total_size;
+	uint16_t conn_sample_size = 0;
+	uint16_t i = 0;
+	void *tmp_data;
+
+	__ASSERT(dt && dt->dev, "device tree info or sensing dev is invalid");
+	sensor_ctx = dt->dev->data;
+	__ASSERT(sensor_ctx, "sensing sensor context is invalid");
+
+	sample_size = sensor_ctx->register_info->sample_size;
+	for (i = 0; i < dt->reporter_num; i++) {
+		conn_sample_size += get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	/* total memory to be alloctead for a sensor according to sensor device tree:
+	 * 1) size of struct sensing_sensor *
+	 * 2) sample data point to struct sensing_sensor->data_buf
+	 * 3) size of struct sensing_connection* for sensor connection to its reporter
+	 * 4) reporter sample size to be stored in connection data
+	 */
+	total_size = sizeof(*sensor) + sample_size +
+		     dt->reporter_num * sizeof(*sensor->conns) + conn_sample_size;
+
+	/* total size for different sensor maybe different, for example:
+	 * there's no reporeter for physical sensor, so no connection memory is needed
+	 * reporter num of each virtual sensor may also different, so connection memory is also
+	 * varied, so here malloc is a must for different sensor.
+	 */
+	sensor = malloc(total_size);
```

Comment:
for each sensor, data size, sensor reporter count, connection count and connection sample size cannot be finalized through device tree define, so malloc is needed to allocate memory for each sensor

Reply:
Malloc is only used by a few samples currently (11 by my count).  It's okay to provide malloc as an option for the sensor subsystem, but there should also be an option to create a fixed pool of connections.

The fixed pool could be configured using Kconfig options.  This is something than can be added in a later PR, but I think it should get prioritized so that the sensor subsystem can be used by more applications.

Reply:
Yes, understand, we tried to avoid malloc, this is the only one we used. 

as the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/55389#discussion_r1161464643, malloc is more flexible and efficiency for memory usages for more applications especially when applications want dynamically open sensors which we don't know at build time. 

Can we keep this malloc in this PR,  and think if have another more better options, I think kconfig with fixed pool is a good candidate.

Reply:
I'm okay with this PR only supporting `malloc`.  But adding Kconfig options to allocate a fixed buffer should be priority for future work.

Reply:
Non-blocking comment regarding the malloc support.

You should statically allocate `struct sensing_sensor` because you always need one entry per sensor found in the devicetree.  The malloc should be limited to the sample buffer.

This will allow you to consolidate the `sensing_sensor` and `sensing_dt_info` structures into a single structure.

Reply:
besides struct sensing_sensor, there're three more: 1. sensor data sample size, 2. connections, 3. samples of connection client size to be mallocated. In our design, we malloc these three together with struct sensing_sensor to manage sensor memory together.